cni/pkg/ebpf/app/ambient_redirect.bpf.c

```diff
@@ -0,0 +1,277 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_endian.h>
+
+#include "ambient_redirect.h"
+
+#define dbg(fmt, ...)                               \
+    ({                                              \
+        bpf_printk("[Debug]: " fmt, ##__VA_ARGS__); \
+    })
+
+
+/* This is an one line Array to store current ztunnel info
+|--------------------------------------------------------------|
+|  0          | Val(ztunnel info: ifindex, mac address)        |
+|--------------------------------------------------------------|
+|  0          |     6, 00:11:22:33:44:55                       |
+|--------------------------------------------------------------|
+*/
+
+struct {
+        __uint(type, BPF_MAP_TYPE_ARRAY);
+        __uint(max_entries, 1);
+        __type(key, __u32);
+        __type(value, struct ztunnel_info);
+        __uint(pinning, LIBBPF_PIN_BY_NAME);
+} ztunnel_info SEC(".maps");
+
+/* This is a hash map to store current application info
+|--------------------------------------------------------------------------|
+|  key(pod IP)         |        Val(app info: ifindex, mac address)        |
+|--------------------------------------------------------------------------|
+|  10.101.1.2          |        2, 00:11:22:33:44:aa                       |
+|--------------------------------------------------------------------------|
+|  10.101.2.2          |        3, 00:11:22:33:44:bb                       |
+|--------------------------------------------------------------------------|
+*/
```

Comment:
I don't really understand why there's an eBPF map for every single pod here - it should be possible to obtain ifindex and mac of the intended dest dynamically, eliminating the need to maintain an eBPF map with every single pod.

Is it an attempt at optimization? I don't see the need for it otherwise, it adds complexity to the CNI layer and creates the potential for us maintaining a kernel data structure of undeterminable size, which I don't like.

Reply:
you mean do the lookup when packet comes in ebpf program itself?  do not think this will perform well

Reply:
Yes. Info is already in the sk_buf. 


It's not clear to me if the performance of getting what's already in the sk_buf is slower than updating our own map from userspace and doing a lookup - but the latter is much simpler for a variety of reasons and avoid us introducing a map with an unknowable upper size bound.

Reply:
the info is not in sk_buf actually.. eg, ztunnel need to find info for app, not himself.

Reply:
using __skbuf, you can find the device/ifidx/mac of the source and dest IPs of the packet - if you can do that for the dest IP in the ztunnel egress prog, you don't need the pod map anymore.

Reply:
The routing table map used here is not necessary and creates extra work/statetracking in CNI userspace, but we can rework that later - and I think in the medium/long term we need to think more carefully about our kernelspace/userspace interface anyway, as a followup.